File: zig/src/day01.zig
const std = @import("std");
const Allocator = std.mem.Allocator;
const expect = std.testing.expect;

fn Pair(comptime T1: type, comptime T2: type) type {
    return struct {
        first: T1,
        second: T2,
    };
}

fn read_data(allocator: Allocator, file_name: []const u8) ![]u8 {
    const file = try std.fs.cwd().openFile(file_name, std.fs.File.OpenFlags{
        .mode = .read_only,
    });
    defer file.close();
    const file_stats = try file.stat();

    const buffer: []u8 = try allocator.alloc(u8, file_stats.size);
    const data_read = try file.read(buffer);

    std.debug.assert(data_read == file_stats.size);

    return buffer;
}

fn parse_data(allocator: Allocator, buffer: []u8) !Pair([]i32, []i32) {
    var splits = std.mem.split(u8, std.mem.trim(u8, buffer, " \n\t\r"), "\n");
    var line_count: usize = 0;
    while (splits.next()) |_| : (line_count += 1) {}
    splits.reset();

    const left = try allocator.alloc(i32, line_count);
    const right = try allocator.alloc(i32, line_count);

    var idx: usize = 0;
    while (splits.next()) |line| : (idx += 1) {
        var nums = std.mem.split(u8, std.mem.trim(u8, line, " \n\t\r"), "   ");
        const left_num = try std.fmt.parseInt(i32, nums.next() orelse "", 10);
        const right_num = try std.fmt.parseInt(i32, nums.next() orelse "", 10);
        left[idx] = left_num;
        right[idx] = right_num;
    }

    return Pair([]i32, []i32){
        .first = left,
        .second = right,
    };
}

fn part1(lists: Pair([]i32, []i32)) u32 {
    var totalDistance: u32 = 0;
    for (0..lists.second.len) |index| {
        totalDistance += @abs(lists.second[index] - lists.first[index]);
    }
    return totalDistance;
}

fn part2(allocator: Allocator, lists: Pair([]i32, []i32)) !i32 {
    var numCounts = std.AutoHashMap(i32, i32).init(allocator);
    defer numCounts.deinit();

    for (lists.second) |item| {
        if (numCounts.get(item)) |count| {
            try numCounts.put(item, count + 1);
        } else {
            try numCounts.put(item, 1);
        }
    }
    var totalSimilarity: i32 = 0;
    for (lists.first) |num| {
        totalSimilarity += (numCounts.get(num) orelse 0) * num;
    }
    return totalSimilarity;
}

pub fn main() !void {
    const allocator = std.heap.page_allocator;

    const data: []u8 = try read_data(allocator, "../../inputs/day01.input");
    defer allocator.free(data);

    const list_pair: Pair([]i32, []i32) = try parse_data(allocator, data);
    std.mem.sort(i32, list_pair.first, {}, std.sort.asc(i32));
    std.mem.sort(i32, list_pair.second, {}, std.sort.asc(i32));

    const totalDistance = part1(list_pair);
    const totalSimilarity = try part2(allocator, list_pair);

    std.debug.assert(totalDistance == 1258579);
    std.debug.assert(totalSimilarity == 23981443);
}

test "day1 test data" {
    const allocator = std.testing.allocator;

    const data: []u8 = try read_data(allocator, "../../tests/day01.test");
    defer allocator.free(data);

    const list_pair: Pair([]i32, []i32) = try parse_data(allocator, data);
    defer {
        allocator.free(list_pair.first);
        allocator.free(list_pair.second);
    }
    std.mem.sort(i32, list_pair.first, {}, std.sort.asc(i32));
    std.mem.sort(i32, list_pair.second, {}, std.sort.asc(i32));

    const totalDistance = part1(list_pair);
    const totalSimilarity = try part2(allocator, list_pair);

    try std.testing.expect(totalDistance == 11);
    try std.testing.expect(totalSimilarity == 31);
}

File: python/day01.py
from typing import List, Tuple
import os
import sys


def read_data(file_name: str) -> Tuple[List[int], List[int]]:
    left_list: List[int] = []
    right_list: List[int] = []
    with open(file_name) as file:
        lines = file.readlines()
        for line in lines:
            left, right = [int(word.strip()) for word in line.strip().split()]
            left_list.append(left)
            right_list.append(right)
    return (left_list, right_list)


def part1(left_list: List[int], right_list: List[int]) -> int:
    total_distance = 0
    for i in range(len(left_list)):
        total_distance += abs(right_list[i]-left_list[i])
    return total_distance


def part2(left_list: List[int], right_list: List[int]) -> int:
    num_count = {}
    for i in range(len(left_list)):
        if (num_count.get(right_list[i])):
            num_count[right_list[i]] += 1
        else:
            num_count[right_list[i]] = 1
    total_similarity = 0
    for num in left_list:
        if (num_count.get(num)):
            total_similarity += num * num_count.get(num)
    return total_similarity


if __name__ == "__main__":
    input_file: str = "/home/mohitjangra/learning/advent_of_code_2024/inputs/day01.input"
    test_file: str = "/home/mohitjangra/learning/advent_of_code_2024/tests/day01.test"

    mode = os.sys.argv[1] if len(os.sys.argv) > 1 else "test"
    if mode.strip().lower() == "input":
        left_list, right_list = read_data(input_file)
    elif mode.strip().lower() == "test":
        left_list, right_list = read_data(test_file)

    left_list = sorted(left_list)
    right_list = sorted(right_list)

    p1_result = part1(left_list, right_list)
    p2_result = part2(left_list, right_list)

    if mode.strip().lower() == "input":
        assert p1_result == 1258579
        assert p2_result == 23981443
    elif mode.strip().lower() == "test":
        assert p1_result == 11
        assert p2_result == 31
    else:
        print(f"Usage: {os.sys.argv[0]} [test|input]", file=sys.stderr)
